sdk/storage/Azure.Storage.DataMovement/src/DataTransferStatus.cs

```diff
@@ -1,66 +1,205 @@
+        ///
+        /// If set to `true`, the transfer has at least one failure item.
+        /// If set to `false`, the transfer currently has no failures.
         /// </summary>
-        InProgress = 2,
+        public bool HasFailedItems { get; internal set; }
 
         /// <summary>
-        /// The Job has been paused. When transfer is paused (e.g. see <see cref="TransferManager.PauseTransferIfRunningAsync(string, System.Threading.CancellationToken)"/>) during the transfer,
-        /// this will be the value.
+        /// Represents if transfer has any skipped items.
         ///
-        /// This status is a resumable state, only
-        /// transfers that failed will be retried when <see cref="TransferManager.StartTransferAsync(StorageResource, StorageResource, DataTransferOptions, CancellationToken)"/>
-        /// with the respective transfer ID to resume.
+        /// If set to `true`, the transfer has at least one item it has skipped.
+        /// If set to `false`, the transfer currently has no items that has been skipped.
+        ///
+        /// It's possible to never have any items skipped if
+        /// <see cref="StorageResourceCreationPreference.SkipIfExists"/> is not enabled in the <see cref="DataTransferOptions.CreationPreference"/>.
+        /// </summary>
+        public bool HasSkippedItems { get; internal set; }
+
+        /// <summary>
+        /// Constructor to set the initial state to <see cref="TransferState.Queued"/> with no failures or skipped items.
         /// </summary>
-        Paused = 3,
+        protected internal DataTransferStatus()
+        {
+            State = TransferState.Queued;
+            HasFailedItems = false;
+            HasSkippedItems = false;
+        }
 
         /// <summary>
-        /// The Job has completed successfully with no failures or skips.
+        /// Constructor to have a custom state, failure state, and skipped state.
         /// </summary>
-        Completed = 4,
+        protected internal DataTransferStatus(TransferState state, bool hasFailureItems, bool hasSkippedItems)
+        {
+            State = state;
+            HasFailedItems = hasFailureItems;
+            HasSkippedItems = hasSkippedItems;
+        }
+
+        internal bool IsCompletedWithFailedItems => State.Equals(TransferState.Completed) && HasFailedItems;
+        internal bool IsCompletedWithSkippedItems => State.Equals(TransferState.Completed) && HasSkippedItems;
 
         /// <summary>
-        /// The Job has been completed with at least one skipped transfer.
+        /// Accordingly update the <see cref="HasFailedItems"/> to true. If already set to true, nothing will happen.
+        ///
+        /// This should only be triggered when a failed item has been seen.
         /// </summary>
-        CompletedWithSkippedTransfers = 5,
+        /// <returns>True if <see cref="HasFailedItems"/> was updated. False otherwise.</returns>
+        internal bool OnFailedItem()
```

Comment:
Yeah we can probably dodge this altogether by using Interlocked. We can just keep a private counter of failed or skipped items and methods like these can just call [`Interlocked.Increment()`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.interlocked.increment). No fumbling over locks. The boolean properties just return if the counter is greater than zero.

Reply:
Actually, do we ever set this to false, or do we only initialize it as false? Because if we don't ever set it normally, do we even need to worry about race conditions? If two threads are racing to set it to true, and none will ever race to set it to false, we get our desired outcome.

Reply:
This doesn't increment anything.

We don't ever set it to false. That's true we won't hit an undesirable race condition. I can remove this for both Failed and Skipped.

Reply:
done.